velox/functions/lib/Re2Functions.cpp

```diff
@@ -1119,120 +1246,185 @@ class PatternStringIterator {
-  // fixed string for patterns of type kSubstring.
-  int32_t fixedPatternEnd = -1;
-  // Count of wildcard character sequences in pattern.
-  size_t numWildcardSequences = 0;
-  // Total number of % characters.
-  size_t anyCharacterWildcardCount = 0;
-  // Total number of _ characters.
-  size_t singleCharacterWildcardCount = 0;
-
+  std::vector<SubPatternKind> subPatternKinds;
+  std::vector<std::pair<size_t, size_t>> subPatternRanges;
   PatternStringIterator iterator{pattern, escapeChar};
 
   // Iterate through the pattern string to collect the stats for the simple
   // patterns that we can optimize.
+  std::ostringstream os;
+  SubPatternKind previousKind;
+  size_t currentSubPatternStart = 0;
+  size_t cursor = 0;
   while (iterator.next()) {
-    const size_t currentStart = iterator.currentStart();
-    if (iterator.isWildcard()) {
-      if (wildcardStart == -1) {
-        wildcardStart = currentStart;
-      }
-
-      if (iterator.isSingleCharWildcard()) {
-        ++singleCharacterWildcardCount;
-      } else {
-        ++anyCharacterWildcardCount;
-      }
-
-      if (!iterator.isPreviousWildcard()) {
-        ++numWildcardSequences;
-      }
-
-      // Mark the end of the fixed pattern.
-      if (fixedPatternStart != -1 && fixedPatternEnd == -1) {
-        fixedPatternEnd = currentStart - 1;
-      }
+    SubPatternKind currentKind;
+    if (iterator.isSingleCharWildcard()) {
+      currentKind = SubPatternKind::kSingleCharWildcard;
+    } else if (iterator.isAnyCharsWildcard()) {
+      currentKind = SubPatternKind::kAnyCharsWildcard;
     } else {
-      // Record the first fixed pattern start.
-      if (fixedPatternStart == -1) {
-        fixedPatternStart = currentStart;
-      } else {
-        // This is not the first fixed pattern, not supported, so fallback.
-        if (iterator.isPreviousWildcard()) {
-          return PatternMetadata{PatternKind::kGeneric, 0};
-        }
-      }
+      currentKind = SubPatternKind::kLiteralString;
     }
-  }
 
-  // The pattern end may not been marked if there is no wildcard char after
-  // pattern start, so we mark it here.
-  if (fixedPatternStart != -1 && fixedPatternEnd == -1) {
-    fixedPatternEnd = patternLength - 1;
-  }
+    // New sub pattern occurs.
+    if (currentKind != previousKind && cursor > 0) {
+      subPatternKinds.push_back(previousKind);
+      subPatternRanges.push_back(
+          {currentSubPatternStart, cursor - currentSubPatternStart});
+      currentSubPatternStart = cursor;
+    }
 
-  // At this point pattern has max of one fixed pattern.
-  // Pattern contains wildcard characters only.
-  if (fixedPatternStart == -1) {
-    if (anyCharacterWildcardCount == 0) {
-      return PatternMetadata{
-          PatternKind::kExactlyN, singleCharacterWildcardCount};
+    // Advance the cursor.
+    cursor++;
+    previousKind = currentKind;
+    os << iterator.current();
+  }
+  std::string unescapedPattern = os.str();
```

Comment:
I wonder if it makes sense to add logic to exit early if the pattern is "obviously" not the one we can optimize for.

This probably doesn't matter much when pattern is constant, but may be important when pattern is non-constant and we end up going through this code for every row. It would be nice to add this use case to the benchmark.

Reply:
Currently we only support optimized pattern which has '%' only at beginning or end of it. So I think we can count the '%' sub-patterns during pattern string parsing, if we has met one '%' sub-pattern, but it is not the first/last sub-pattern, we return kGeneric immediately. How do you think? @mbasmanova

Reply:
This sounds good. Thanks.

Reply:
Added corresponding fast break path.